### ISSUE TITLE
Consider forking the query language as a separate lib

### ISSUE THREAD

**Issue by smarie** [msiemens/tinydb#428]
Hi @msiemens , first of all let me say that I like this lib very much, as it provides an elegant and pythonic API for queries, with a reference file-based implementation that is "good enough" for many projects.

In our team we have an in-house storage mechanism that very much resembles a document store, and where we are implementing queries.

Our architecture currently allows for multiple backend implementations - one of them can be tinydb, another mongodb, etc. However this is not the case for the query API.

I thought first that we could simply reuse the `Query` class from TinyDB, and then let the backend transform this into the actual query language of its choice (e.g. pymongo). However I figured out that your implementation does not keep track of the various operations when we use operators. In other word when we do `Query().a == 1`, we can still read the `_path` containing `('a',)`, but we can not read the "equal" operator with the right-hand-side element `1`.

Therefore I will have to modify your classes so that they "remember" the operations rather than executing them. That way, a backend will be able to "read back" the queries and transform them into another query language (pymongo...).

Would you be interested in having such a feature directly baked-in the `Query` object ? If so, we could even imagine to fork these classes as an independent lib (`pydocquery` or whater name you'll like :) ), so that both tinydb AND other db projects could support the language.

Let me know what you think !

**Comment by msiemens** [comment]
Hey @smarie, that's an interesting question!

> In other word when we do `Query().a == 1`, we can still read the `_path` containing `('a',)`, but we can not read the "equal" operator with the right-hand-side element `1`: they are already gone and integrated in a callable.

Hmm… I think you could use `QueryInstance._hash` to get a representation of your query:

```python
>>> from tinydb.queries import Query
>>> q = Query().some.path > 5
>>> q._hash
('>', ('some', 'path'), 5)
```

This is because TinyDB's query cache needs to be able to tell if two queries are the same or not. This is done by hashing whatever is in `QueryInstance._hash` which in turn is an abstract representation of the original query.

> Would you be interested in having such a feature directly baked-in the `Query` object ? If so, we could even imagine to fork these classes as an independent lib (`pydocquery`, `pydocql`, etc. or whater name you'll like :) ), so that both tinydb AND other db projects could support the language.

That's an interesting idea. Off the top of my head my first thought is that this would make it difficult to add new query operations, especially if other projects are not able to implement new query operations e.g. due to technical constraints.

That being saig, the idea itself of having a generalized abstract query representation is charming. Maybe there could be a different approach of a sort of _Document Query Representation_ standard (kind of an abstract syntax tree for queries) that multiple projects could support. Basically TinyDB could support converting its queries into this standard represention and back. There could be one or multiple projects that provide different interfaces to generate standard query objects. But maybe that's all too complicated. I mean, what good is a standard that nobody uses (or only 1 or 2 projects)?

So basically, the most simple solution would be to use `QueryInstance`'s `_hash` to find out what operation the query represents. But that of course is a dependency on TinyDB implementation details which are subject to change. Having a standard document query representation would be the other extreme, which would involve coordination between projects that implement this standard.

**Comment by smarie** [comment]
Thanks @msiemens for both the short-term workaround and longer-term answer !

Concerning the sort-term, the fact that hash is a TinyDB implementation detail subject to change, is definitely a major concern.

Concerning the longer-term, I agree that creating a standard is not very useful if nobody uses it :) . Another possibility for me is to use an existing, yet non-pythonic standard, as a colleague of mine pointed out: https://pypi.org/project/jsonpath-ng/ . Still, I believe from both what I see in `tinydb` and from my past experience developing a similar lib for callables ([`mini-lambda`](https://smarie.github.io/python-mini-lambda/)), that the python language is intuitive enough to cover 50%, maybe more, of the cases. And that many users do not want to learn a new language just to select all documents that contain a value above a threshold.

I'll continue to think about it and let you know. (you may close the ticket)
Thanks again !
